pkg/combustion/registry.go

```diff
@@ -0,0 +1,173 @@
+
+	data, err := template.Parse(haulerManifestYamlName, haulerManifest, ctx.ImageDefinition.EmbeddedArtifactRegistry)
+	if err != nil {
+		return fmt.Errorf("applying template to %s: %w", haulerManifestYamlName, err)
+	}
+
+	if err := os.WriteFile(haulerManifestYamlFile, []byte(data), fileio.NonExecutablePerms); err != nil {
+		return fmt.Errorf("writing file %s: %w", haulerManifestYamlName, err)
+	}
+
+	return nil
+}
+
+func populateHaulerStore(ctx *image.Context, haulerExecutable string) error {
+	haulerManifestPath := filepath.Join(ctx.BuildDir, haulerManifestYamlName)
+	args := []string{"store", "sync", "--files", haulerManifestPath}
+
+	cmd, registryLog, err := createRegistryCommand(ctx, haulerExecutable, args)
+	if err != nil {
+		return fmt.Errorf("preparing to populate registry store: %w", err)
+	}
+	defer func() {
+		if err = registryLog.Close(); err != nil {
+			zap.S().Warnf("failed to close registry log file properly: %s", err)
+		}
+	}()
+
+	if err = cmd.Run(); err != nil {
+		return fmt.Errorf("populating hauler store: %w: ", err)
+	}
+
+	return nil
+}
+
+func generateRegistryTar(ctx *image.Context, haulerExecutable string) error {
+	haulerTarDest := filepath.Join(ctx.CombustionDir, registryTarName)
+	args := []string{"store", "save", "--filename", haulerTarDest}
+
+	cmd, registryLog, err := createRegistryCommand(ctx, haulerExecutable, args)
+	if err != nil {
+		return fmt.Errorf("preparing to generate registry tar: %w", err)
+	}
+	defer func() {
+		if err = registryLog.Close(); err != nil {
+			zap.S().Warnf("failed to close registry log file properly: %s", err)
+		}
+	}()
+
+	if err = cmd.Run(); err != nil {
+		return fmt.Errorf("creating registry tar: %w: ", err)
+	}
+
+	return nil
+}
+
+func copyHaulerBinary(ctx *image.Context, haulerExecutable string) error {
+	destinationDir := filepath.Join(ctx.CombustionDir, "hauler")
+
+	err := fileio.CopyFile(haulerExecutable, destinationDir, fileio.ExecutablePerms)
+	if err != nil {
+		return fmt.Errorf("copying hauler binary to combustion dir: %w", err)
+	}
+
+	return nil
+}
+
+func writeRegistryScript(ctx *image.Context) (string, error) {
+	values := struct {
+		EmbeddedRegistryTar string
+	}{
+		EmbeddedRegistryTar: registryTarName,
+	}
+
+	data, err := template.Parse(registryScriptName, registryScript, &values)
+	if err != nil {
+		return "", fmt.Errorf("parsing registry script template: %w", err)
+	}
+
+	filename := filepath.Join(ctx.CombustionDir, registryScriptName)
+	err = os.WriteFile(filename, []byte(data), fileio.ExecutablePerms)
+	if err != nil {
+		return "", fmt.Errorf("writing registry script: %w", err)
+	}
+
+	return registryScriptName, nil
+}
+
+func createRegistryCommand(ctx *image.Context, commandName string, args []string) (*exec.Cmd, *os.File, error) {
+	fullLogFilename := filepath.Join(ctx.BuildDir, registryLogFileName)
+	logFile, err := os.OpenFile(fullLogFilename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, fileio.NonExecutablePerms)
```

Comment:
I'm a little bit conflicted on this one. Not sure how large the output is but if we reuse the same file then it is harder for us to track down the issues once something fails. What do you think?

Reply:
I was also debating the same thing, but ultimately decided that it might be too much to have 2 different log files, but honestly I’m fine with that if it’s gonna be easier in the long run when it comes to debugging.

Reply:
Feel free to include an example of what it currently looks like so we can decide based on that.